### PR TITLE
Write event log if wrong image is used

### PR DESCRIPTION
**Pull request by chinhtranvan**
This PR adds event logging for cases where an incorrect image is used. It covers scenarios where an image suited for admintools is used for a vclusterops deployment and vice versa.

### CODE REVIEW

**Comment by chinhtranvan** [comment]
I test my change with my manifest for two case
If I use v2 image for admintools
```
Events:
  Type     Reason      Age                From                Message
  ----     ------      ----               ----                -------
  Warning  WrongImage  19s (x4 over 92s)  verticadb-operator  Failed to choose an image suited for admintools
```
and If I use v1 image for vcluster
```  
Events:
  Type     Reason      Age                From                Message
  ----     ------      ----               ----                -------
  Warning  WrongImage  15s (x3 over 31s)  verticadb-operator  Failed to choose an image suited for vclusterops
```

**Comment by chinhtranvan** [comment]
@spilchen  I tested my changes using my manifest for two scenarios: first, if I use the v2 image for admintools
```
Events:
  Type     Reason      Age                From                Message
  ----     ------      ----               ----                -------
  Warning  WrongImage  19s (x4 over 92s)  verticadb-operator  Failed to choose an image suited for admintools
```
and If I use v1 image for vcluster
```  
Events:
  Type     Reason      Age                From                Message
  ----     ------      ----               ----                -------
  Warning  WrongImage  15s (x3 over 31s)  verticadb-operator  Failed to choose an image suited for vclusterops
```

File: pkg/controllers/vdb/image_reconciler.go [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+		c.Log.Info("No pods running") // No running pod.  This isn't an error, it just means no vertica is running
+		return ctrl.Result{}, nil
+	}
+	stdout, _, _ := c.PRunner.ExecInPod(ctx, pf.name, names.ServerContainer,
+		"bash", "-c", "which admintools",
+	)
+	res := strings.TrimSpace(stdout) // remove leading and trailing whitespaces
+	if vmeta.UseVClusterOps(c.Vdb.Annotations) {
+		if res == admintoolsBin {
+			c.VRec.Eventf(c.Vdb, corev1.EventTypeWarning, events.WrongImage, "Failed to choose an image suited for vclusterops")
```
Comment (spilchen) [inline pkg/controllers/vdb/image_reconciler.go:71]: Suggest a different error message:
```suggestion
			c.VRec.Eventf(c.Vdb, corev1.EventTypeWarning, events.WrongImage, "Image cannot be used for vclusterops deployments")
```

File: pkg/controllers/vdb/image_reconciler.go [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+			return ctrl.Result{}, fmt.Errorf("failed to choose an image suited: the image %s is an admintools style of deployment "+
+				",cannot be used for vcluster", c.Vdb.Spec.Image)
+		}
+	} else {
+		if res == "" {
+			c.VRec.Eventf(c.Vdb, corev1.EventTypeWarning, events.WrongImage, "Failed to choose an image suited for admintools")
```
Comment (spilchen) [inline pkg/controllers/vdb/image_reconciler.go:77]: Suggest a different error message:
```suggestion
			c.VRec.Eventf(c.Vdb, corev1.EventTypeWarning, events.WrongImage, "Image cannot be used for admintools deployments")
```

File: pkg/controllers/vdb/image_reconciler.go [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+	}
+	stdout, _, _ := c.PRunner.ExecInPod(ctx, pf.name, names.ServerContainer,
+		"bash", "-c", "which admintools",
+	)
+	res := strings.TrimSpace(stdout) // remove leading and trailing whitespaces
+	if vmeta.UseVClusterOps(c.Vdb.Annotations) {
+		if res == admintoolsBin {
+			c.VRec.Eventf(c.Vdb, corev1.EventTypeWarning, events.WrongImage, "Failed to choose an image suited for vclusterops")
+			return ctrl.Result{}, fmt.Errorf("failed to choose an image suited: the image %s is an admintools style of deployment "+
+				",cannot be used for vcluster", c.Vdb.Spec.Image)
```
Comment (spilchen) [inline pkg/controllers/vdb/image_reconciler.go:73]: Suggest a different error message:
```suggestion
			return ctrl.Result{}, fmt.Errorf("image %s is meant for admintools style of deployments and cannot be used for vclusterops.",
				c.Vdb.Spec.Image)
```

File: pkg/controllers/vdb/image_reconciler.go [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+		}
+	} else {
+		if res == "" {
+			c.VRec.Eventf(c.Vdb, corev1.EventTypeWarning, events.WrongImage, "Failed to choose an image suited for admintools")
+			return ctrl.Result{}, fmt.Errorf("failed to choose an image suited: the image %s is a vcluster style of deployment "+
+				",cannot be used for admintools", c.Vdb.Spec.Image)
```
Comment (spilchen) [inline pkg/controllers/vdb/image_reconciler.go:79]: Suggest a different error message:
```suggestion
			return ctrl.Result{}, fmt.Errorf("image %s is meant for vclusterops style of deployments and cannot be used for admintools.",
				c.Vdb.Spec.Image)
```

File: pkg/controllers/vdb/verticadb_controller.go [hunk 5]
```diff
@@ -180,6 +180,8 @@ func (r *VerticaDBReconciler) constructActors(log logr.Logger, vdb *vapi.Vertica
 		// Handle setting up the pod security context. This picks the
 		// UID/fsGroup that we will run with.
 		MakePodSecurityReconciler(r, log, vdb),
+		// Handle wrong images for the operator
+		MakeImageCheckReconciler(r, log, vdb, prunner, pfacts),
```
Comment (spilchen) [inline pkg/controllers/vdb/verticadb_controller.go:184]: I'm not sure if this is the correct spot. For create/revive, we haven't even created the pods yet. But we also need to do the check early for upgrade code paths. So, we probably need it in a couple of places. I think we need to do the check in the same spots that we capture the version. I would suggest rather than add a new reconciler, lets add your logic to the version reconciler (`version_reconciler.go). Then we checking it in the correct spot.

**Review by spilchen — COMMENTED** [review]
Can we add an e2e test? We have a leg that is vcluster ops only (e2e-leg-6). It would be a big test. We would just attempt to create a db using the vclusterops image and assert that we see the event message.

File: pkg/controllers/vdb/verticadb_controller.go (hunk 5 above)
Comment (roypaulin) [reply pkg/controllers/vdb/verticadb_controller.go:184]: Then should we call it `ImageVersionReconciler`? where we check both the version and the deployment type.

File: pkg/controllers/vdb/offlineupgrade_reconciler.go [hunk 6]
```diff
@@ -249,7 +249,7 @@ func (o *OfflineUpgradeReconciler) checkVersion(ctx context.Context) (ctrl.Resul
 	}
 
 	const EnforceUpgradePath = true
-	vr := MakeVersionReconciler(o.VRec, o.Log, o.Vdb, o.PRunner, o.PFacts, EnforceUpgradePath)
+	vr := MakeVersionImageReconciler(o.VRec, o.Log, o.Vdb, o.PRunner, o.PFacts, EnforceUpgradePath)
```
Comment (roypaulin) [inline pkg/controllers/vdb/offlineupgrade_reconciler.go:252]: `MakeImageVersionReconciler` is better

File: pkg/controllers/vdb/onlineupgrade_reconciler.go [hunk 7]
```diff
@@ -454,11 +454,11 @@ func (o *OnlineUpgradeReconciler) checkVersion(ctx context.Context, sts *appsv1.
 	}
 
 	const EnforceUpgradePath = true
-	a := MakeVersionReconciler(o.VRec, o.Log, o.Vdb, o.PRunner, o.PFacts, EnforceUpgradePath)
+	a := MakeVersionImageReconciler(o.VRec, o.Log, o.Vdb, o.PRunner, o.PFacts, EnforceUpgradePath)
 
 	// We use a custom lookup function to only find pods for the subcluster we
 	// are working on.
-	vr := a.(*VersionReconciler)
+	vr := a.(*VersionImageReconciler)
```
Comment (roypaulin) [inline pkg/controllers/vdb/onlineupgrade_reconciler.go:461]: `ImageVersionReconciler` is better.

File: pkg/controllers/vdb/version_reconciler_test.go [hunk 8]
```diff
@@ -33,6 +33,9 @@ var _ = Describe("k8s/version_reconcile", func() {
 
 	It("should update annotations in vdb since they differ", func() {
 		vdb := vapi.MakeVDB()
+		vdb.ObjectMeta.Annotations = map[string]string{
```
Comment (roypaulin) [inline pkg/controllers/vdb/version_reconciler_test.go:36]: Why did you add this?

File: pkg/controllers/vdb/version_reconciler.go [hunk 9]
```diff
@@ -69,6 +71,20 @@ func (v *VersionReconciler) Reconcile(ctx context.Context, _ *ctrl.Request) (ctr
 		return ctrl.Result{Requeue: true}, nil
 	}
 
+	if vmeta.UseVClusterOps(v.Vdb.Annotations) {
```
Comment (roypaulin) [inline pkg/controllers/vdb/version_reconciler.go:74]: We should add a test case for this part

Comment (chinhtranvan) [reply pkg/controllers/vdb/version_reconciler.go:74]: we have e2e leg 6 for this part

File: pkg/controllers/vdb/version_reconciler_test.go (hunk 8 above)
Comment (chinhtranvan) [reply pkg/controllers/vdb/version_reconciler_test.go:36]: For this test, 'vertica-k8s:latest' is deployed using the Vcluster style, not the AT style. If we use the default False annotation with Vcluster deployment, this test will fail. I have set the annotation to true so they can pass verification for the image. Is there any way to deploy 'vertica-k8s:latest' using the AT style?

**Comment by chinhtranvan** [comment]
> Can we add an e2e test? We have a leg that is vcluster ops only (e2e-leg-6). It would be a big test. We would just attempt to create a db using the vclusterops image and assert that we see the event message.

@spilchen i already address all your comments. I just wonder for the e2e test, Should we move it to e2e leg1? When e2e-leg-6 is vcluster only (with v2 image deployment and vclusterOps annotation is True), I wonder we can’t hit the event wrong image. I try to update vclusterOps annotation , and it still can’t see the event.

File: pkg/controllers/vdb/version_reconciler.go (hunk 9 above)
Comment (roypaulin) [reply pkg/controllers/vdb/version_reconciler.go:75]: Yes sorry by test case I meant unit test

File: pkg/controllers/vdb/imageversion_reconciler_test.go [hunk 10]
```diff
@@ -88,12 +91,39 @@ vertica(v11.1.0) built by @re-docker2 from tag@releases/VER_10_1_RELEASE_BUILD_1
...
+		r := MakeImageVersionReconciler(vdbRec, logger, vdb, fpr, &pfacts, true)
 		Expect(r.Reconcile(ctx, &ctrl.Request{})).Should(Equal(ctrl.Result{Requeue: true}))
 
 		// Ensure we didn't update the vdb
 		fetchVdb := &vapi.VerticaDB{}
 		Expect(k8sClient.Get(ctx, vapi.MakeVDBName(), fetchVdb)).Should(Succeed())
 		Expect(fetchVdb.ObjectMeta.Annotations[vmeta.VersionAnnotation]).Should(Equal(OrigVersion))
 	})
+
+	It("should fail the reconclier if we use wrong image", func() {
```
Comment (roypaulin) [inline pkg/controllers/vdb/imageversion_reconciler_test.go:103]: The test should cover wrong image error for both admintools and vclusterops

Comment (chinhtranvan) [reply pkg/controllers/vdb/imageversion_reconciler_test.go:103]: Thank for your comments, I'm trying to find a way to build an image using AT style to verity wrong image for vclusterops.

Comment (roypaulin) [reply pkg/controllers/vdb/imageversion_reconciler_test.go:103]: unit tests and e2e tests are different. your should have unit test that cover both cases regardless of e2e test

Comment (roypaulin) [reply pkg/controllers/vdb/imageversion_reconciler_test.go:103]: No image build needed. Only go code. In one case you have the vclusterops flag unset and a Podfact with admintoolsExists=false, on the other case  vclusterops flag set and a Podfact with admintoolsExists=true.

Comment (chinhtranvan) [reply pkg/controllers/vdb/imageversion_reconciler_test.go:103]: Done!

File: pkg/controllers/vdb/podfacts.go [hunk 11]
```diff
@@ -426,6 +431,8 @@ func (p *PodFacts) genGatherScript(vdb *vapi.VerticaDB, pf *PodFact) string {
 		df --block-size=1 --output=size %s | tail -1
 		echo -n 'localDataAvail: '
 		df --block-size=1 --output=avail %s | tail -1
+		echo -n 'admintoolsExists: '
+        which admintools &> /dev/null && echo true || echo false
```
Comment (spilchen) [inline pkg/controllers/vdb/podfacts.go:435]: Can you fix the alignment with this line?

File: pkg/controllers/vdb/imageversion_reconciler_test.go [hunk 12]
```diff
@@ -88,12 +91,41 @@ vertica(v11.1.0) built by @re-docker2 from tag@releases/VER_10_1_RELEASE_BUILD_1
...
+		defer test.DeleteVDB(ctx, k8sClient, vdb)
+		test.CreatePods(ctx, k8sClient, vdb, test.AllPodsRunning)
+		defer test.DeletePods(ctx, k8sClient, vdb)
+
+		fpr := &cmds.FakePodRunner{}
+		pfacts := MakePodFacts(vdbRec, fpr)
+		Expect(pfacts.Collect(ctx, vdb)).Should(Succeed())
+
+		r := MakeImageVersionReconciler(vdbRec, logger, vdb, fpr, &pfacts, true)
+		// both the vclusterops annotation and admintools exit are false
```
Comment (spilchen) [inline pkg/controllers/vdb/imageversion_reconciler_test.go:116]: minor: spelling
```suggestion
		// both the vclusterops annotation and admintoolsExists are false
```

**Review by spilchen — COMMENTED** [review]
Looks good. Just some very minor comments.

File: pkg/controllers/vdb/podfacts.go (hunk 11 above)
Comment (chinhtranvan) [reply pkg/controllers/vdb/podfacts.go:435]: It's weird, I didn't have this one on my local machine

Comment (chinhtranvan) [reply pkg/controllers/vdb/podfacts.go]: ah tabs issues, just fix it!

**Review by spilchen — APPROVED** [review]

**Comment by spilchen** [comment]
I saw your new testcase failed in the latest e2e run. We should have a closer look at that failure to see if there isn't any timing issue.

**Review by roypaulin — APPROVED** [review]

**Comment by chinhtranvan** [comment]
> I saw your new testcase failed in the latest e2e run. We should have a closer look at that failure to see if there isn't any timing issue.

I examined the log and identified why this test is unstable; the reason is a timing issue. When we set up the vdb at step 15, it starts creating the database before we enable the annotation vclusterop at step 17, leading to the successful createDB. In the log, it is evident that we attempted to stop the database later using API HTTP and NMA endpoints (via vclusterOPs), but we failed at the stop_db step. By synchronizing steps 15 and 17, we can enable the vcluster annotation at the right time and avoid the timing issue.

**Comment by spilchen** [comment]
Thanks for following up with the test.
